service/src/main/java/com/commercetools/pspadapter/payone/PaymentHandler.java

```diff
@@ -0,0 +1,111 @@
+
+import com.commercetools.pspadapter.payone.domain.ctp.CommercetoolsQueryExecutor;
+import com.commercetools.pspadapter.payone.domain.ctp.PaymentWithCartLike;
+import com.commercetools.pspadapter.payone.domain.ctp.exceptions.NoCartLikeFoundException;
+import io.sphere.sdk.client.ErrorResponseException;
+import io.sphere.sdk.client.NotFoundException;
+import io.sphere.sdk.http.HttpStatusCode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import java.util.ConcurrentModificationException;
+import java.util.concurrent.CompletionException;
+
+import static com.commercetools.pspadapter.tenant.TenantLoggerUtil.createLoggerName;
+import static io.sphere.sdk.http.HttpStatusCode.INTERNAL_SERVER_ERROR_500;
+import static java.lang.String.format;
+import static java.util.Optional.ofNullable;
+
+public class PaymentHandler {
+
+    /**
+     * How many times to retry if {@link ConcurrentModificationException} happens.
+     */
+    private static final int RETRIES_LIMIT = 20;
+    private static final int RETRY_DELAY = 100; // msec
+
+    private final String payoneInterfaceName;
+
+    private final CommercetoolsQueryExecutor commercetoolsQueryExecutor;
+    private final PaymentDispatcher paymentDispatcher;
+
+    private final Logger logger;
+
+    public PaymentHandler(String payoneInterfaceName, String tenantName,
+                          CommercetoolsQueryExecutor commercetoolsQueryExecutor, PaymentDispatcher paymentDispatcher) {
+        this.payoneInterfaceName = payoneInterfaceName;
+
+        this.commercetoolsQueryExecutor = commercetoolsQueryExecutor;
+        this.paymentDispatcher = paymentDispatcher;
+
+        this.logger = LoggerFactory.getLogger(createLoggerName(this.getClass(), tenantName));
+    }
+
+    /**
+     * Tries to handle the payment with the provided ID.
+     *
+     * @param paymentId identifies the payment to be processed
+     * @return the result of handling the payment
+     */
+    public PaymentHandleResult handlePayment(@Nonnull final String paymentId) {
+        try {
+            for (int i = 0; i < RETRIES_LIMIT; i++) {
+                try {
+                    final PaymentWithCartLike payment = commercetoolsQueryExecutor.getPaymentWithCartLike(paymentId);
+                    String paymentInterface = payment.getPayment().getPaymentMethodInfo().getPaymentInterface();
+                    if (!payoneInterfaceName.equals(paymentInterface)) {
+                        logger.warn("Wrong payment interface name: expected [{}], found [{}]", payoneInterfaceName, paymentInterface);
+                        return new PaymentHandleResult(HttpStatusCode.BAD_REQUEST_400);
+                    }
+
+                    paymentDispatcher.dispatchPayment(payment);
+                    return new PaymentHandleResult(HttpStatusCode.OK_200);
+                } catch (final ConcurrentModificationException cme) {
+                    Thread.sleep(RETRY_DELAY);
```

Comment:
Does the caller get any information why it failed. Will the last exception be logged after you cached it here?